org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/http/HttpClient.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * SonarLint for Eclipse
+ * Copyright (C) 2015-2024 SonarSource SA
+ * sonarlint@sonarsource.com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.core.internal.http;
+
+import java.net.URI;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.NoSuchAlgorithmException;
+import java.time.Duration;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import org.eclipse.jdt.annotation.Nullable;
+import org.sonarlint.eclipse.core.SonarLintLogger;
+import org.sonarsource.sonarlint.core.rpc.protocol.backend.initialize.HttpConfigurationDto;
+
+/**
+ *  Very bare bones implementation of a HTTP client that can work in the context of SonarLint both with a configurable
+ *  key and trust store. Additionally, timeouts can be provided that will used the same way as SLCORE would use them.
+ */
+public class HttpClient {
```

Comment:
It seems like reinventing the wheel a lot. Couldn't you leverage the existing HTTP client in SLCORE?

Reply:
Sadly no :cry:

Before 10.0, it was still available in the `org.sonarsource.sonarlint.core.clientapi.SonarLintBackend` but only to open HTTP connections to a specific connection. Since 10.0 it was completely moved to the part out of process, not being available to the clients anymore.

I thought about sparking a conversation with the Sonar Solution squad about it, but in the end decided against it for the following reasons:
- having an HTTP client as API would enable (malicious) third parties to plug into that and use it to trigger calls on behalf of SonarLint (Core), e.g. when SonarLint is the only plug-in configured with HTTP/SSL configurations and others are not
- in general, no client should make any HTTP requests as everything they need is either in SLCORE or provided by the IDE, here this is a solution as Eclipse doesn't provide what is needed (and probably never will be) - information about the Release Notes as well as newly released versions
- the implementation is as bare-bones as it gets, I was even thinking in the first place of dropping the `SSLContext` related logic but keeping it otherwise might have rendered this feature useless - the HTTP/SSL configuration provided by users is there for a reason in the first place

Reply:
I was not thinking about using the HTTP client in SLE directly, but instead creating a dedicated service in SLCORE. This could be, perhaps, reused in other flavors.